overlord/snapstate/handlers.go

```diff
@@ -310,20 +336,115 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
 	return ts, err
 }
 
-func (m *SnapManager) installPrereqs(t *state.Task, base string, prereq []string, userID int, tm timings.Measurer) error {
+// updates a prerequisite, if it's not providing a content interface that a plug expects it to
+func updatePrereqIfOutdated(t *state.Task, snapName string, contentAttrs []string, userID int, flags Flags) (*state.TaskSet, error) {
+	st := t.State()
+
+	if len(contentAttrs) == 0 {
+		return nil, nil
+	}
+
+	if ok, err := shouldSkipToAvoidConflict(t, snapName); err != nil {
+		return nil, err
+	} else if ok {
+		return nil, nil
+	}
+
+	// check if the default provider can provide the content it's meant to
+	ok, err := hasAllContentAttrs(st, snapName, contentAttrs)
+	if err != nil {
+		return nil, err
+	}
+
+	if !ok {
+		// default provider is missing some content tags (likely outdated) so update it
+		ts, err := Update(st, snapName, nil, userID, flags)
+		if err != nil {
+			if _, ok := err.(*ChangeConflictError); ok {
+				return nil, &state.Retry{After: prerequisitesRetryTimeout, Reason: fmt.Sprintf("update failed due to conflict: %s", err)}
+			} else {
+				t.Logf("installed prerequisite '%s' is missing content. Failed to update: %s", snapName, err)
+			}
+			return nil, nil
+		}
+
+		return ts, nil
+	}
+
+	return nil, nil
+}
+
+// Checks for conflicting tasks. Returns true if the operation should be skipped. The error
+// can be a state.Retry if the operation should be retried later.
+func shouldSkipToAvoidConflict(task *state.Task, snapName string) (bool, error) {
+	tasks, err := inFlightLinkSnapTasks(task.State(), snapName)
+	if err != nil {
+		return false, err
+	}
+
+	for _, otherTask := range tasks {
+		if otherTask.ID() == task.ID() {
+			continue
+		}
+
+		// it's in the same change, so the snap is already going to be installed
+		if otherTask.Change().ID() == task.Change().ID() {
+			return true, nil
+		}
+
+		// it's not in the same change, so retry to avoid conflicting changes to the snap
+		return true, &state.Retry{
+			After:  prerequisitesRetryTimeout,
+			Reason: fmt.Sprintf("conflicting changes on snap %q by task %q", snapName, otherTask.Kind()),
+		}
+	}
+
+	return false, nil
+}
+
+// Checks if the snap has slots with "content" attributes matching the
+// ones that the snap being installed requires
+func hasAllContentAttrs(st *state.State, snapName string, requiredContentAttrs []string) (bool, error) {
+	providedContentAttrs := make(map[string]bool)
+	repo := ifacerepo.Get(st)
+
+	for _, slot := range repo.Slots(snapName) {
+		val, ok := slot.Lookup("content")
+		if !ok {
+			continue
+		}
+
+		contentAttr, ok := val.(string)
```

Comment:
there's no test hitting this

Reply:
Added tests to fully cover `hasAllContentAttrs`